### PR TITLE
[OpenAI] EmbeddingItems is able to take float array from customer's json object

### PR DESCRIPTION
**Pull request by mssfang**
close: https://github.com/Azure/azure-sdk-for-java/issues/41159

It is a nice feature to add.

It is possible user wants to serialize or deserialize the `EmbeddingItem` from another place but not from our SDK.

### CODE REVIEW

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java [hunk 1]
```diff
@@ -19,15 +20,37 @@
 @Immutable
 public final class EmbeddingItem implements JsonSerializable<EmbeddingItem> {
 
+    /*
+     * List of embeddings value for the input prompt. These represent a measurement of the
+     * vector-based relatedness of the provided input.
+     */
+    @Generated
+    private final List<Double> embedding;
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:28]: Why do we use `List<Double>` instead of `List<Float>` here?

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java [hunk 2]
```diff
@@ -19,15 +20,37 @@
...
+
     private final String embeddingBase64;
 
+    /**
+     * Creates an instance of EmbeddingItem class.
+     *
+     * @param embeddingBase64 the embedding value to set.
+     * @param promptIndex the promptIndex value to set.
+     */
+    private EmbeddingItem(String embeddingBase64, List<Double> embedding, int promptIndex) {
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:38]: Do we need this private constructor?

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java [hunk 3]
```diff
@@ -79,36 +107,28 @@ public JsonWriter toJson(JsonWriter jsonWriter) throws IOException {
...
             while (reader.nextToken() != JsonToken.END_OBJECT) {
                 String fieldName = reader.getFieldName();
                 reader.nextToken();
                 if ("embedding".equals(fieldName)) {
-                    embedding = reader.getString();
+                    JsonToken jsonToken = reader.currentToken();
+                    if (jsonToken == JsonToken.STRING) {
+                        embedding = reader.getString();
+                    } else if (jsonToken == JsonToken.START_ARRAY) {
+                        embeddingInDouble = reader.readArray(JsonReader::getDouble);
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:120]: Can we convert this to base64 string here instead of having the private constructor?

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java (hunk 1 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:28]: It is default to Double when generated.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java (hunk 2 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:38]: A new private constructor will be generated if we remove this private constructor. It is for passing the typespec validation.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java (hunk 1 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:28]: image

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java (hunk 2 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:38]: image

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java (hunk 3 above)
Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:126]: We have to keep on private constructor to skip the TypeSpec validaton. From this snapshot you can see actually List<Double> is what generated but not String. So Keep editing the private constructor and keep generated annotation for List<DouBle> 
image

I have added a helper method to convert the List<Float> to base64 encoded string. So when user pass a list of floats, they will have base64encoded string as well.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java [hunk 4]
```diff
@@ -9,25 +9,51 @@
...
      * Get the embedding property: List of embeddings value for the input prompt. These represent a measurement of the
      * vector-based relatedness of the provided input.
      *
      * @return the embedding value.
      */
     public List<Float> getEmbedding() {
+        if (embedding != null) {
+            return embedding.stream().map(Double::floatValue).collect(Collectors.toList());
+        }
         return convertBase64ToFloatList(embeddingBase64);
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:57]: We don't want to call convert every time `getEmbedding` method is called. 

```suggestion
        embedding = convertBase64ToFloatList(embeddingBase64);
        return embedding;
```

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java [hunk 5]
```diff
@@ -37,7 +63,10 @@ public List<Float> getEmbedding() {
      * @return the embedding base64 encoded string.
      */
     public String getEmbeddingAsString() {
-        return embeddingBase64;
+        if (embeddingBase64 != null) {
+            return embeddingBase64;
+        }
+        return convertFloatListToBase64(embedding.stream().map(Double::floatValue).collect(Collectors.toList()));
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:69]: Same here - we should convert only once.

```suggestion
        embeddingBase64 = convertFloatListToBase64(embedding.stream().map(Double::floatValue).collect(Collectors.toList()));
        return embeddingBase64;
```

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java [hunk 6]
```diff
@@ -62,8 +91,13 @@ public int getPromptIndex() {
     @Override
     public JsonWriter toJson(JsonWriter jsonWriter) throws IOException {
         jsonWriter.writeStartObject();
-        jsonWriter.writeStringField("embedding", this.embeddingBase64);
-        jsonWriter.writeIntField("index", this.promptIndex);
+        jsonWriter.writeStringField("object", "embedding");
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:94]: Why is this added?

Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:94]: (1) OpenAI's payload has this field "object": "embedding"

image

[API Reference - OpenAI API](https://platform.openai.com/docs/api-reference/embeddings/object)



(2) Generated code has used the same pattern:

image

[[OpenAI-Inference] Inference SDK support for Azure OpenAI `v2024-07-01-preview` service release by jpalvarezl · Pull Request #41210 · Azure/azure-sdk-for-java (github.com)](https://github.com/Azure/azure-sdk-for-java/pull/41210/files#diff-1158b0e826e3bd10ddd27567b1fa563dfda502395a7a52d02a95f56575a7b0b2R25)

Comment (srnagar) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java:96]: Is this a new field introduced in the latest version? It was not there in previous versions.

Comment (mssfang) [reply sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/models/EmbeddingItem.java]: new field removed.

File: sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/EmbeddingsUtils.java [hunk 7]
```diff
@@ -41,4 +42,18 @@ public static List<Float> convertBase64ToFloatList(String embedding) {
         }
         return floatList;
     }
+
+    // This method converts a list of floats to a base64 string
```
Comment (srnagar) [inline sdk/openai/azure-ai-openai/src/main/java/com/azure/ai/openai/implementation/EmbeddingsUtils.java:46]: This should be a javadoc - `/** .... */`

**Review by srnagar — APPROVED** [review]
